src/lightning/data/cache/writer.py

```diff
@@ -0,0 +1,273 @@
+
+        if self._data_format is None:
+            self._data_format = data_format
+        elif self._data_format != data_format:
+            raise Exception(
+                f"The data format changed between items. Found {data_format} instead of {self._data_format}."
+            )
+
+        if self._data_spec is None:
+            self._data_spec = data_spec
+        elif self._data_spec != data_spec:
+            raise Exception(f"The data format changed between items. Found {data_spec} instead of {self._data_spec}.")
+
+        head = np.array(sizes, np.uint32).tobytes()
+        body = b"".join(data)
+        return head + body
+
+    def _serialize(self, item, sizes, data) -> bytes:
+        for serializer_name, serializer in self._serializers.items():
+            if serializer.can_serialize(item):
+                serialized_item = serializer.serialize(item)
+                data.append(serialized_item)
+                sizes.append(len(serialized_item))
+                return serializer_name
+        raise ValueError(f"The provided item isn't serializable. Found {item}")
+
+    def _create_chunk(self, filename: str) -> bytes:
+        """Create a binary chunk from all the binarized items."""
+        num_items = np.uint32(len(self._serialized_items))
+        sizes = list(map(len, self._serialized_items))
+        offsets = np.array([0] + sizes).cumsum().astype(np.uint32)
+        offsets += len(num_items.tobytes()) + len(offsets.tobytes())
+        sample_data = b"".join(self._serialized_items)
+        data = num_items.tobytes() + offsets.tobytes() + sample_data
+        offsets = offsets.tolist()
+        mapping = {}
+        for i in range(len(self._indexes)):
+            mapping[self._indexes[i]] = [offsets[i], offsets[i + 1]]
+
+        assert len(mapping) == len(self._indexes)
+        assert (self._indexes[-1] - self._indexes[0] + 1) == len(self._serialized_items)
+
+        chunk_info = {
+            "samples": len(self._serialized_items),
+            "filename": filename,
+            "mapping": mapping,
+            "interval": [self._indexes[0], self._indexes[-1] + 1],
+        }
+
+        self._chunks_info.append(chunk_info)
+
+        return data
+
+    def write_chunk(self):
+        """Write a chunk to the filesystem."""
+        if self._compression:
+            filename = f"chunk-{self.rank}-{self._chunk_id}.{self._compression}.bin"
+        else:
+            filename = f"chunk-{self.rank}-{self._chunk_id}.bin"
+        self.write_file(self._create_chunk(filename), filename)
+
+    @property
+    def available_serializers(self):
+        return self._serializers
+
+    def reset(self) -> None:
+        """Reset the writer to handle the next chunk."""
+        self._serialized_items = []
+        self._indexes = []
+        self._current_chunk_size = 0
+
+    def __setitem__(self, index, items: any):
+        """Store an item to a chunk.
+
+        The index needs to be provided in order.
```

Comment:
Can you clarify the requirement? Also I see the assert below but we should properly raise a proper exception instead.

Reply:
This is handled by the samplers automatically. This ensures we can map an index to a shard from an interval.

Reply:
Added the comment.